### ISSUE TITLE
Adding support for mix inverters

### ISSUE THREAD

**Issue by Dom3442** [vtraveller/gowatt#1]
Hi first up thanks for creating and publishing this.

I have a mix inverter and think I have worked out how to alter your method to make it work for mix inverters.  Basically if you change every mention of spa in code to mix then it still works.  I have produced an altered version of your code that works with my system so thanks for publishing this code and giving me a good start on this.  The login also needs to be altered as rawGetEicDevices() does not contain the deviceSn for me (but you can get that from rawGetDevices() instead on my system (a SPH3600).

**Comment by vtraveller** [comment]
That's great. Many thanks for deciphering the part I couldn't. I'll update it.

**Comment by vtraveller** [comment]
I'll upload a ref.txt shortly.  Can you confirm this assumption?  In particular the case of the various mix references.

```
panel/spa/getSPAStatusData -> panel/mix/getMIXStatusData
panel/spa/getSPABatChart -> panel/mix/getMIXBatChart
panel/spa/getSPAEnergyDayChart -> panel/mix/getMIXEnergyDataChart
panel/spa/getSPAEnergyMonthChart -> panel/mix/getMIXEnergyMonthChart
panel/spa/getSPAEnergyYearChart -> panel/mix/getMIXEnergyYearChart
panel/spa/getSPAEnergyTotalChart -> panel/mix/getMIXEnergyTotalChart
spaSn -> mixSn
spa.status.normal -> mix.status.normal
"panelType":"mix"
```

Many thanks.  I'm thinking of using rawGetDevices():
 "deviceTypeName":"spa"

To absolve the need for any SPA or MIX references in the API.

**Comment by vtraveller** [comment]
Normalized code uploaded.  I can't guarantee it'll work with a mix, but its following the same pattern as SPA.

**Comment by Dom3442** [comment]
Thanks.  Tried new version and it works apart from your setRuleBatteryFirst function.  My inverter requires an additional parameter to enable AC charging.  So schedule settings for mix inverters should be.

    schedule_settings = [
      '100',                        # Charging power %
      str(amount),                  # Stop charging when above SoC %
     str('1' if enable else '0'),  # Enable AC charging - Enabled/Disabled (1 = Enabled)
      str(startHour), '00',         # Schedule 1 - Start time
      str(endHour), '00',           # Schedule 1 - End time
      str('1' if enable else '0'),  # Schedule 1 - Enabled/Disabled (1 = Enabled)
      '00','00',                    # Schedule 2 - Start time
      '00','00',                    # Schedule 2 - End time
      '0',                          # Schedule 2 - Enabled/Disabled (1 = Enabled)
      '00','00',                    # Schedule 3 - Start time
      '00','00',                    # Schedule 3 - End time
      '0'                           # Schedule 3 - Enabled/Disabled (1 = Enabled)
    ]

**Comment by vtraveller** [comment]
Added mix for the battery rule, as per your advice.  I squeezed it into the existing structure as I didn't want things to drift too far ... at least until Growatt made some radically annoying change.

Also chucked in some helpers after someone asked about solar rates and the units of measure.  The most interesting function being the getGridRate() which is a calculation based on available rates of the other functions.  I discovered that it can pull excess from the battery randomly that sort of vanishes (10W), so I've assumed that's a grid export ... but it may just be echoes in the DC Clamps.

I'll leave the issue open as I anticipate further divergence between spa and mix.
